### PR TITLE
add organization ID check from JWT token for internal rules

### LINKED ISSUE #76: Implement organization ID check for internal rules
**Description:**
Internal rules are going to be added to content-service, which will make smart-proxy pull them and start serving them (namely on Single Rule endpoint), but these internal rules are way more sensitive than external rules, therefore only people from Red Hat organization should be able to access this information from smart-proxy

**Proposed solution**
- re-implement organization whitelist for single rule endpoint

**Acceptance criteria**
- users from the Red Hat organization are able to access internal rules via single rule endpoints
- users outside of Red Hat org are forbidden to access internal rules

### PR DESCRIPTION
**Pull request by Bee-lee**
# Description
This adds a security check for internal rules based on organization ID

Fixes #76 
Blocks https://github.com/RedHatInsights/insights-content-service/pull/97
Blocked by https://github.com/RedHatInsights/e2e-deploy/pull/1923

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update
- Refactor (refactoring code, removing useless files)
- Unit tests (no changes in the code)
- Documentation update

## Testing steps
Modified requests to contain diffferent org_ids to test manually
+ I had some timeout issues locally, so let's see what travis has to say

### CODE REVIEW

File: content/parsing.go [hunk 1]
```diff
@@ -122,3 +120,39 @@ func timeParse(value string) (time.Time, error) {
...
+// IsRuleInternal tries to look for the word "internal" in the ruleID / rule module,
+// because it's currently not specified anywhere on it's own
+// TODO: add field indicating restricted/internal status to one of Rule structs in content-service
+func IsRuleInternal(ruleID types.RuleID) bool {
+	isInternal := false
+	splitRuleID := strings.Split(string(ruleID), ".")
+
+	for _, ruleIDPart := range splitRuleID {
+		if ruleIDPart == internalRuleStr {
+			isInternal = true
```
Comment (Sergey1011010) [inline content/parsing.go:152]: I'd just return true here, but it's a nitpick :)

File: server/server.go [hunk 2]
```diff
@@ -519,12 +520,56 @@ func (server HTTPServer) singleRuleEndpoint(writer http.ResponseWriter, request
 		return
 	}
 
+	if rule.Internal == true {
```
Comment (Sergey1011010) [inline server/server.go:523]: I think just `if rule.Internal {` is slightly better, but it's a nitpick again :)

File: server/server.go [hunk 3]
```diff
@@ -519,12 +520,56 @@ func (server HTTPServer) singleRuleEndpoint(writer http.ResponseWriter, request
...
 		log.Error().Err(err).Msg(responseDataError)
 	}
 }
 
+// checkInternalRulePermissions checks if organizations for internal rules are enabled
+// if so, retrieves the org_id from request/token and returns whether that ID is on the list
+// of allowed organizations to access internal rules
+func (server HTTPServer) checkInternalRulePermissions(writer http.ResponseWriter, request *http.Request) bool {
+	if !server.Config.EnableInternalRulesOrganizations || !server.Config.Auth {
+		return true
```
Comment (Sergey1011010) [inline server/server.go:542]: Does it mean that if internal rules are disabled, user would still get them or I'm missing something?

Comment (Sergey1011010) [reply server/server.go:542]: Ah, got it, it's about restricting internal rules to some organizations from the list, not the rules at all :)

File: server/server.go [hunk 4]
```diff
@@ -519,12 +520,56 @@ func (server HTTPServer) singleRuleEndpoint(writer http.ResponseWriter, request
...
+	authToken, err := server.GetAuthToken(request)
+	if err != nil {
+		handleServerError(writer, err)
+		return false
+	}
+
+	requestOrgID := int(authToken.Internal.OrgID)
+
+	log.Info().Msgf("Checking internal rule permissions for Organization ID: %v", requestOrgID)
+	for _, allowedID := range server.Config.InternalRulesOrganizations {
```
Comment (Sergey1011010) [inline server/server.go:556]: Btw, how many rules internal rules organizations do we expect?

File: server/server_test.go [hunk 5]
```diff
@@ -184,4 +254,62 @@ func TestHTTPServer_ReportEndpoint(t *testing.T) {
...
+// TestDisabledInternalOrganizations tests accessing internal rule when EnableInternalRulesOrganizations = false
+func TestDisabledInternalOrganizations(t *testing.T) {
+	helpers.RunTestWithTimeout(t, func(t *testing.T) {
+
+		loadMockRuleContentDir(RuleContentInternal1)
+
+		_, err := content.GetRuleContent(internalTestRuleModule)
+		helpers.FailOnError(t, err)
+
+		helpers.AssertAPIRequest(t, &serverConfigJWT, &helpers.DefaultServicesConfig, nil, &helpers.APIRequest{
```
Comment (Sergey1011010) [inline server/server_test.go:306]: Btw, these tests could be refactored by using `t.Run("sub test name", func...`

File: server/server.go (hunk 3 above)
Comment (tisnik) [reply server/server.go:542]: yeah, but in reality, the filter will be disabled for most time AFAIK

File: server/server.go (hunk 4 above)
Comment (tisnik) [reply server/server.go:556]: possibly it will be disabled (I mean the filter)

Comment (tisnik) [reply server/server.go:556]: (I know it is not the right answer though :)

File: server/server_test.go (hunk 5 above)
Comment (Bee-lee) [reply server/server_test.go]: oh right, thanks! done

File: server/server.go (hunk 4 above)
Comment (Bee-lee) [reply server/server.go:556]: @Sergey1011010 @tisnik exactly, from zero to I dont know :)

File: server/server.go (hunk 2 above)
Comment (Bee-lee) [reply server/server.go]: thanks:)

File: server/handlers.go [hunk 6]
```diff
@@ -59,6 +59,15 @@ func (server HTTPServer) getContentForRule(writer http.ResponseWriter, request *
 		return
 	}
 
+	// check for internal rule permissions
+	if internal := content.IsRuleInternal(ruleID); internal == true {
```
Comment (joselsegura) [inline server/handlers.go:63]: `internal == true` => `internal`

Just a nitpick

File: server/handlers.go [hunk 7]
```diff
@@ -59,6 +59,15 @@ func (server HTTPServer) getContentForRule(writer http.ResponseWriter, request *
...
+		ok := server.checkInternalRulePermissions(writer, request)
+		if ok != true {
```
Comment (joselsegura) [inline server/handlers.go:65]: `ok != true` => `!ok`

File: server/server_test.go [hunk 8]
```diff
@@ -101,13 +129,54 @@ var (
...
+					Description: testdata.RuleErrorKey1.Description,
+					Impact:      testdata.ImpactIntToStr[testdata.RuleErrorKey1.Impact],
+					Likelihood:  testdata.RuleErrorKey1.Likelihood,
+					PublishDate: testdata.RuleErrorKey1.PublishDate.UTC().Format(time.RFC3339),
+					Tags:        testdata.RuleErrorKey1.Tags,
+					Status:      "active",
+				},
+			},
+		},
+	}
```
Comment (joselsegura) [inline server/server_test.go:158]: Move this definition to `testdata` package in `insights-results-aggregator-data` repository, but it can be done later

**Pull request merged**